ikuzo/service/x/ead/oaipmh.go

```diff
@@ -0,0 +1,161 @@
+package ead
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/delving/hub3/hub3/ead"
+	"github.com/delving/hub3/ikuzo/service/x/oaipmh"
+	"github.com/kiivihal/goharvest/oai"
+	"github.com/rs/zerolog/log"
+)
+
+type EADHarvester struct {
+	s     *Service
+	OrgID string
+}
+
+func NewEADHarvester(s *Service) (EADHarvester, error) {
+	if s == nil {
+		return EADHarvester{}, fmt.Errorf("nil value not allowed for ead.Service")
+	}
+
+	return EADHarvester{s: s}, nil
+}
+
+func (e *EADHarvester) ProcessEadFromOai(r *oai.Response) {
+	if r.Request.Verb != oaipmh.VerbListRecords {
+		log.Warn().Str("verb", r.Request.Verb).Msg("verb is not supported for getting ead records")
+		return
+	}
+
+	for _, record := range r.ListRecords.Records {
+		record := record
+
+		log.Info().
+			Str("identifier", record.Header.Identifier).
+			Strs("set", record.Header.SetSpec).
+			Str("lastModified", record.Header.DateStamp).
+			Str("recordStatus", record.Header.Status).
+			Msg("ead file entry")
+
+		if err := e.processRecord(&record); err != nil {
+			log.Error().
+				Str("identifier", record.Header.Identifier).
+				Err(err).
+				Msg("unable to process ead record")
+		}
+	}
+}
+
+func (e *EADHarvester) processRecord(record *oai.Record) error {
+	archiveID := record.Header.Identifier
+	body := record.Metadata.Body
+
+	if len(body) == 0 {
+		return fmt.Errorf("metadata.Body cannot be empty")
+	}
+
+	r := bytes.NewReader(body)
+
+	_, meta, err := e.s.SaveEAD(r, int64(len(body)), archiveID, e.OrgID)
+	if err != nil {
+		if errors.Is(err, ErrTaskAlreadySubmitted) {
+			e.s.M.IncAlreadyQueued()
+			return nil
+		}
+
+		e.s.M.IncFailed()
+
+		return err
+	}
+
+	t, err := e.s.NewTask(&meta)
+	if err != nil {
+		e.s.M.IncAlreadyQueued()
+		return err
+	}
+
+	log.Info().
+		Str("orgID", t.Meta.OrgID).
+		Str("datasetID", t.Meta.OrgID).
+		Str("taskID", t.ID).
+		Msg("submitted task for indexing via OAI-PMH")
+
+	return nil
+}
+
+type MetsHarvester struct {
+	c     *ead.DaoClient
+	OrgID string
+}
+
+func NewMetsHarvest(c *ead.DaoClient) (MetsHarvester, error) {
+	if c == nil {
+		return MetsHarvester{}, fmt.Errorf("nil value not allowed for DaoClient")
```

Comment:
If a nil value as ead.DaoClient is not allowed. Wouldn't it be better to return nil as MetsHarvester?